src/graph/ChcGraph.cc

```diff
@@ -599,4 +601,126 @@ std::unique_ptr<ChcDirectedHyperGraph> ChcDirectedHyperGraph::makeEmpty(Logic &
+    }
+
+    PTRef transitionRelation = logic.mkOr(std::move(transitionRelationComponent));
+    std::string name;
+    for (auto vertex : vertices) {
+        name += logic.getSymName(vertex);
+    }
+
+    SymRef newRef = logic.declareFun(name, logic.getSort_bool(), args);
+    predicates.addRepresentation(newRef, std::move(vars));
+
+    // KB: Changing input and output edges to and from the loop in the graph with new state and loc variables
+    //     Redirecting edges to the newly created vertice
+    allEdges = getEdges();
+    for (auto edge : allEdges) {
+        if (std::find(vertices.begin(), vertices.end(), getSource(edge)) != vertices.end()) {
+            PTRef badStates = locationVars.at(getSource(edge));
+            auto oldEdgeVars = getVariablesFromEdge(logic, *this, edge);
+            std::unordered_map<PTRef, PTRef, PTRefHash> subMap;
+            for (uint i = 0; i < oldEdgeVars.stateVars.size(); i++) {
+                subMap.insert(std::make_pair(oldEdgeVars.stateVars[i], argVars[{getSource(edge), i}]));
+            }
+            updateEdgeSource(edge, newRef);
+            PTRef label = utils.varSubstitute(logic.mkAnd(getEdgeLabel(edge), badStates), subMap);
+
+            updateEdgeLabel(edge, InterpretedFla{label});
+        }
+        if (std::find(vertices.begin(), vertices.end(), getTarget(edge)) != vertices.end()) {
+
+            PTRef initialStates = [&]() -> PTRef {
+                vec<PTRef> negatedLocations;
+                negatedLocations.capacity(locationVars.size());
+                for (auto && entry : locationVars) {
+                    if (std::get<0>(entry) != getTarget(edge)) {
+                        negatedLocations.push(logic.mkNot(timeMachine.sendVarThroughTime(entry.second, 1)));
+                    }
+                }
+                return logic.mkAnd(std::move(negatedLocations));
+            }();
+            PTRef destination = timeMachine.sendVarThroughTime(locationVars.at(getTarget(edge)), 1);
+
+            auto oldEdgeVars = getVariablesFromEdge(logic, *this, edge);
+            std::unordered_map<PTRef, PTRef, PTRefHash> subMap;
+            for (uint i = 0; i < oldEdgeVars.nextStateVars.size(); i++) {
+                subMap.insert(std::make_pair(oldEdgeVars.nextStateVars[i],
+                                             timeMachine.sendVarThroughTime(argVars[{getTarget(edge), i}], 1)));
+            }
+            updateEdgeTarget(edge, newRef);
+            PTRef label = utils.varSubstitute(getEdgeLabel(edge), subMap);
+
+            updateEdgeLabel(edge, InterpretedFla{rewriteMaxArityClassic(
+                                            logic, logic.mkAnd(logic.mkAnd(initialStates, destination), label))});
+        }
+    }
+
+    // KB: Creating new self-looping edge (which represents all of the transitions in the outer loop)
+    newEdge(newRef, newRef, InterpretedFla{transitionRelation});
+    return {newRef, locationVars, argVars};
```

Comment:
I am still concerned about the case where there are additional edges between the nodes of the loop.
You will end up with multiple self-loops on the new node, which I think is not want you want.
You want to have single self-loop in the end, no?

Reply:
This should be taken care of at the lines 642-647
All of the possible edges between the nodes inside the loop are removed

Reply:
OK, I see. I agree that that should work correctly.